crates/volta-core/src/tool/yarn/fetch.rs

```diff
@@ -0,0 +1,150 @@
+//! Provides fetcher for Yarn distributions
+
+use std::fs::{rename, File};
+use std::path::{Path, PathBuf};
+
+use super::super::download_tool_error;
+use crate::error::ErrorDetails;
+use crate::fs::{create_staging_dir, create_staging_file, ensure_containing_dir_exists};
+use crate::hook::ToolHooks;
+use crate::path;
+use crate::style::{progress_bar, tool_version};
+use crate::tool::{self, Yarn};
+use crate::version::VersionSpec;
+use archive::{Archive, Tarball};
+use cfg_if::cfg_if;
+use log::debug;
+use semver::Version;
+use volta_fail::{Fallible, ResultExt};
+
+cfg_if! {
+    if #[cfg(feature = "mock-network")] {
+        fn public_yarn_server_root() -> String {
+            mockito::SERVER_URL.to_string()
+        }
+    } else {
+        fn public_yarn_server_root() -> String {
+            "https://github.com/yarnpkg/yarn/releases/download".to_string()
+        }
+    }
+}
+
+pub fn fetch(version: &Version, hooks: Option<&ToolHooks<Yarn>>) -> Fallible<()> {
+    let yarn_dir = path::yarn_inventory_dir()?;
+    let cache_file = yarn_dir.join(path::yarn_distro_file_name(&version.to_string()));
+
+    if let Some(archive) = load_cached_distro(&cache_file) {
```

Comment:
This could probably be refactored/broken apart along similar lines to the similar functionality I identified in one of the other packages (the Node one IIRC).

Reply:
Good call, I'll do that here and in `package/fetch.rs` as they all follow a roughly similar pattern.

Reply:
Hmm, actually looking at this it's slightly different behavior, because the after-unpack operation in the non-cached case refers to a variable that was only created in the non-cached branch: we only create a `NamedTempFile` to download into if we're actually downloading, if we're unpacking from the cache then it's already there and we don't have to worry about it.

We could return an `Option<NamedTempFile>` from the match, but then it starts to be a question of whether that is cleaner to read or not. Thoughts?

Reply:
Ah, so you end up with something like this, then:

```rust
let (archive, staging) = match load_cached_distro(&cache_file) {
    Some(archive) => {
        debug!(
            "Loading {} from cached archive at '{}'",
            tool_version("yarn", &version),
            cache_file.display(),
        );
        (archive, None)
    }
    None => {
        let staging = create_staging_file()?;
        let remote_url = determine_remote_url(&version, hooks)?;
        let archive = fetch_remote_distro(&version, &remote_url, staging.path())?;

        (archive, Some(staging))
    }
};

unpack_archive(archive, version)?;

if let Some(staging_file) = staging {
    ensure_containing_dir_exists(&cache_file)?;
    staging_file
        .persist(cache_file)
        .with_context(|_| ErrorDetails::PersistInventoryError { tool: "Yarn".into() })?;
}

Ok(())
```

🤔 I could go either way on whether that's clearer. It still has the benefit of separating out the disjoint operations, but at the cost of adding a few lines and possibly some lack of clarity. Totally your call.

Reply:
That's actually not as unclear as I was picturing, so I'll do that. It also has the benefit of being consistent with the `package::fetch` implementation.